third_party/move/move-bytecode-verifier/src/signature_v2.rs

```diff
@@ -21,9 +22,271 @@ use std::{
 };
 use typed_arena::Arena;
 
+/***************************************************************************************************
+ * Pseudo Metering
+ *
+ **************************************************************************************************/
```

Comment:
I'm wondering how this relates to the existing metering approach: https://github.com/aptos-labs/aptos-core/blob/main/third_party/move/move-bytecode-verifier/src/meter.rs. What's the reason this rolls up its own?

Reply:
@wrwg This is because the actual metering logic here is extremely simple -- you keep subtracting some size from a variable representing your budget, and it is the only thing this has in common with meter.rs. I found it awkward to only use trait Meter as a simple counter, while not needing the other APIs (e.g. scope, transfer etc.).

I can see the point of having some common meter construct for the whole bytecode verifier though. Maybe that's your original intention with meter.rs?

Reply:
The meter type is already used in various components (absint-based but also type safety). Please check out the existing examples. Since its an `&mut impl Meter` when passed through, it supports zero abstraction cost if needed. Also it already has the error logic you are re-implementing here. You wont need scope et. al; this is for composing functions costs in a module (and currently has no effect IIRC).

Perhaps the concern here is that then you need to align the metering units with the remaining components? This is indeed a certain challenge. Typically this is done by assigning a cost as in

```
  const TYPE_NODE_COST: u128 = 30;
  ...
  fn charge_ty(&mut self, meter: &mut impl Meter, ty: &SignatureToken) -> PartialVMResult<()> {
        meter.add_items(
            Scope::Function,
            TYPE_NODE_COST,
            ty.preorder_traversal().count(),
        )
    }
```

This is a bit more cumbersome to calibrate but it makes it easier to give a reasonable bound to execution as a whole (plus brings us closer to a gas model).

Reply:
I think I'm starting to see the bigger picture here. Is my understanding correct that you want us to have a unified meter interface (`trait Meter`) that is used by all metering components? Let's also discuss our long-term vision for metering.

> Perhaps the concern here is that then you need to align the metering units with the remaining components?

This should not be a major problem. Although I do have one question: do you think we should have a shared meter budget for all verifier passes? (My current implementation sets a separate budget for the signature checker.)

Reply:
> I think I'm starting to see the bigger picture here. Is my understanding correct that you want us to have a unified meter interface (`trait Meter`) that is used by all metering components? Let's also discuss our long-term vision for metering.
> 
> > Perhaps the concern here is that then you need to align the metering units with the remaining components?
> 
> This should not be a major problem. Although I do have one question: do you think we should have a shared meter budget for all verifier passes? (My current implementation sets a separate budget for the signature checker.)

Yes, that is the goal and already happening in multiple places in the existing code.

The goal is to have some global cost limit which we can set (similar as in gas). This requires callibration towards actual execution cost, which is currently a bit awkward and happening over the multiple constants in the code. The hope is that we can transition this eventually into gas charging. One `impl Meter` could then connect to the gas meter infra in the future..

Reply:
Sounds good. I'll switch over to `trait Meter`.

> Although I do have one question: do you think we should have a shared meter budget for all verifier passes? (My current implementation sets a separate budget for the signature checker.)

Still, what your take on this? My feeling is that if we want proper metering eventually then we should just have a global budget?